Source/DafnyCore/Resolver/ModuleResolver.cs

```diff
@@ -2861,11 +2861,7 @@ void MarkSCCAsNotSupportingEquality() {
             return;  // we are done
           }
           foreach (var arg in ctor.Formals) {
-            var anotherIndDt = arg.Type.AsIndDatatype;
-            if (arg.IsGhost ||
-                (anotherIndDt != null && anotherIndDt.EqualitySupport == IndDatatypeDecl.ES.Never) ||
-                arg.Type.IsCoDatatype ||
-                arg.Type.IsArrowType) {
+            if (arg.IsGhost || SurelyNeverSupportEquality(arg.Type)) {
```

Comment:
We need this type traversal to be recursive, otherwise soundness issue 5972c.dfy